In [1]:
import pandas as pd
import json

pd.set_option("display.max_columns", 100)
pd.set_option("display.max_rows", 25)

In [2]:
! source ../conf/local.env && curl --request GET \
  --url "https://api.openalex.org/sources?per_page=25&api_key=${NACSOS_OPENALEX__API_KEY}"  > sample_25.json

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1033k  100 1033k    0     0   696k      0  0:00:01  0:00:01 --:--:--  696k


In [3]:
with open("sample_25.json", "r") as file:
    txt = json.load(file)
txt.keys()

dict_keys(['meta', 'results', 'group_by'])

In [4]:
df = pd.DataFrame(txt["results"])
df

,id,issn_l,issn,display_name,host_organization,host_organization_name,host_organization_lineage,works_count,oa_works_count,cited_by_count,summary_stats,is_oa,is_in_doaj,is_in_doaj_since_year,is_high_oa_rate,is_high_oa_rate_since_year,is_in_scielo,is_ojs,is_core,oa_flip_year,first_publication_year,last_publication_year,ids,homepage_url,apc_prices,apc_usd,country_code,societies,alternate_titles,type,topics,topic_share,counts_by_year,works_api_url,updated_date,created_date
0,https://openalex.org/S7407052672,None,[],NIFS,None,None,[],38965892,38965892,0,"{'2yr_mean_citedness': 0.0, 'h_index': 0, 'i10...",False,False,None,False,NaN,False,False,False,NaN,2024.0,2026.0,{'openalex': 'https://openalex.org/S7407052672...,None,None,None,None,[],[],repository,"[{'id': 'https://openalex.org/T12157', 'displa...","[{'id': 'https://openalex.org/T11920', 'displa...","[{'year': 2026, 'works_count': 10025756, 'oa_w...",https://api.openalex.org/works?filter=primary_...,2026-05-02T10:02:47,2025-10-10T00:00:00
1,https://openalex.org/S7407053449,None,[],DiSSCo,None,None,[],12065707,12065707,2,"{'2yr_mean_citedness': 1.6575903923408714e-07,...",False,False,None,False,NaN,False,False,False,NaN,2024.0,2026.0,{'openalex': 'https://openalex.org/S7407053449...,None,None,None,None,[],[],repository,"[{'id': 'https://openalex.org/T12157', 'displa...","[{'id': 'https://openalex.org/T14254', 'displa...","[{'year': 2026, 'works_count': 4708281, 'oa_wo...",https://api.openalex.org/works?filter=primary_...,2026-06-12T10:06:09,2025-10-10T00:00:00
2,https://openalex.org/S4306400562,None,None,Zenodo (CERN European Organization for Nuclear...,https://openalex.org/I67311998,European Organization for Nuclear Research,[https://openalex.org/I67311998],11999065,11999065,943452,"{'2yr_mean_citedness': 0.022547823753480706, '...",True,False,None,True,NaN,False,False,False,NaN,1543.0,2026.0,{'openalex': 'https://openalex.org/S4306400562...,https://zenodo.org,[],None,None,[],[],repository,"[{'id': 'https://openalex.org/T13958', 'displa...","[{'id': 'https://openalex.org/T14200', 'displa...","[{'year': 9999, 'works_count': 77, 'oa_works_c...",https://api.openalex.org/works?filter=primary_...,2026-06-12T10:06:09,2016-06-24T00:00:00
3,https://openalex.org/S4377196282,None,None,Figshare,https://openalex.org/I4210132348,Figshare (United Kingdom),[https://openalex.org/I4210132348],10037808,10037807,508563,"{'2yr_mean_citedness': 0.01251073531620561, 'h...",False,False,None,False,NaN,False,False,False,NaN,2006.0,2021.0,{'openalex': 'https://openalex.org/S4377196282...,None,[],None,None,[],[],repository,"[{'id': 'https://openalex.org/T10015', 'displa...","[{'id': 'https://openalex.org/T10451', 'displa...","[{'year': 2922, 'works_count': 1, 'oa_works_co...",https://api.openalex.org/works?filter=primary_...,2026-06-12T10:06:09,2016-06-24T00:00:00
4,https://openalex.org/S4306525036,None,None,PubMed,https://openalex.org/I1299303238,National Institutes of Health,[https://openalex.org/I1299303238],7891092,675049,46057364,"{'2yr_mean_citedness': 0.5067873303167421, 'h_...",False,False,None,False,NaN,False,False,False,NaN,1400.0,2026.0,{'openalex': 'https://openalex.org/S4306525036...,https://pubmed.ncbi.nlm.nih.gov,[],None,US,[],[],repository,"[{'id': 'https://openalex.org/T14134', 'displa...","[{'id': 'https://openalex.org/T13365', 'displa...","[{'year': 2026, 'works_count': 11666, 'oa_work...",https://api.openalex.org/works?filter=primary_...,2026-06-12T10:06:09,2016-06-24T00:00:00
5,https://openalex.org/S4210203682,0366-4457,[0366-4457],Bulletin of Miscellaneous Information (Royal G...,None,None,[],7181824,97313,550213,"{'2yr_mean_citedness': 0.029509276539686026, '...",False,False,None,False,NaN,False,False,False,NaN,0.0,2026.0,{'openalex': 'https://openalex.org/S4210203682...,http://www.archive.org/details/mobot3175300225...,[],None,GB,[],"[Bulletin of miscellaneous Information, Kew]",journal,"[{'id': 'https://openalex.org/T10346', 'displa...","[{'id': 'https://openalex.org/T13436', 'd

In [5]:
ids = pd.DataFrame.from_dict(df.ids.to_dict(), orient="index")
ids

,openalex,issn_l,issn,mag,wikidata
0,https://openalex.org/S7407052672,None,[],7407052672,None
1,https://openalex.org/S7407053449,None,[],7407053449,None
2,https://openalex.org/S4306400562,None,None,4306400562,https://www.wikidata.org/wiki/Q22661177
3,https://openalex.org/S4377196282,None,None,4377196282,None
4,https://openalex.org/S4306525036,None,None,4306525036,https://www.wikidata.org/entity/Q180686
5,https://openalex.org/S4210203682,0366-4457,[0366-4457],4210203682,https://www.wikidata.org/entity/Q5735532
6,https://openalex.org/S4398183272,None,None,4398183272,None
7,https://openalex.org/S7407053304,None,[],7407053304,None
8,https://openalex.org/S7407053410,None,[],7407053410,None
9,https://openalex.org/S7407056385,None,None,7407056385,None


In [6]:
df.merge(ids, left_index=True, right_index=True)[["id", "openalex", "issn_x", "issn_y", "issn_l_x", "issn_l_y", "mag"]]

,id,openalex,issn_x,issn_y,issn_l_x,issn_l_y,mag
0,https://openalex.org/S7407052672,https://openalex.org/S7407052672,[],[],None,None,7407052672
1,https://openalex.org/S7407053449,https://openalex.org/S7407053449,[],[],None,None,7407053449
2,https://openalex.org/S4306400562,https://openalex.org/S4306400562,None,None,None,None,4306400562
3,https://openalex.org/S4377196282,https://openalex.org/S4377196282,None,None,None,None,4377196282
4,https://openalex.org/S4306525036,https://openalex.org/S4306525036,None,None,None,None,4306525036
5,https://openalex.org/S4210203682,https://openalex.org/S4210203682,[0366-4457],[0366-4457],0366-4457,0366-4457,4210203682
6,https://openalex.org/S4398183272,https://openalex.org/S4398183272,None,None,None,None,4398183272
7,https://openalex.org/S7407053304,https://openalex.org/S7407053304,[],[],None,None,7407053304
8,https://openalex.org/S7407053410,https://openalex.org/S7407053410,[],[],None,None,7407053410
9,https://openalex.org/S7407056385,https://openalex.org/S7407056385,None,None,None,None,7407056385


In [7]:
! source ../conf/local.env && curl --request GET \
  --url "https://api.openalex.org/sources?sample=25&select=ids%2Cdisplay_name%2Chost_organization%2Chost_organization_name%2Chost_organization_lineage&api_key=${NACSOS_OPENALEX__API_KEY}" 

{"meta": {"count": 25, "db_response_time_ms": 38, "page": 1, "per_page": 25, "groups_count": null, "x_query": {"oql": "sources\nsample 25\nreturn ids, name, publisher, host organization name, host organization lineage", "oqo": {"get_rows": "sources", "sample": 25, "select": ["ids", "display_name", "host_organization", "host_organization_name", "host_organization_lineage"]}, "url": "/sources?select=ids,display_name,host_organization,host_organization_name,host_organization_lineage&sample=25"}, "cost_usd": 0.0001}, "results": [{"ids": {"openalex": "https://openalex.org/S4210175487", "issn_l": "1517-9265", "issn": ["1517-9265", "2175-0742"], "mag": "4210175487", "wikidata": null}, "display_name": "Fronteiras", "host_organization": "https://openalex.org/P4310312188", "host_organization_name": "Federal University of Grande Dourados", "host_organization_lineage": ["https://openalex.org/P4310312188"]}, {"ids": {"openalex": "https://openalex.org/S4306484997", "issn_l": null, "issn": null, "mag

## Work with full list of sources

In [8]:
oa = pd.read_json("openalex_sources.json")
print(oa.shape)
oa.head()

(283281, 5)


,ids,display_name,host_organization,host_organization_name,host_organization_lineage
0,{'openalex': 'https://openalex.org/S7407052672...,NIFS,None,None,[]
1,{'openalex': 'https://openalex.org/S7407053449...,DiSSCo,None,None,[]
2,{'openalex': 'https://openalex.org/S4306400562...,Zenodo (CERN European Organization for Nuclear...,https://openalex.org/I67311998,European Organization for Nuclear Research,[https://openalex.org/I67311998]
3,{'openalex': 'https://openalex.org/S4377196282...,Figshare,https://openalex.org/I4210132348,Figshare (United Kingdom),[https://openalex.org/I4210132348]
4,{'openalex': 'https://openalex.org/S4306525036...,PubMed,https://openalex.org/I1299303238,National Institutes of Health,[https://openalex.org/I1299303238]


In [9]:
ids = pd.json_normalize(oa["ids"]).add_prefix("id_")
oa = pd.concat([oa.drop(columns=["ids"]), ids], axis=1)
oa.head()

,display_name,host_organization,host_organization_name,host_organization_lineage,id_openalex,id_issn_l,id_issn,id_mag,id_wikidata
0,NIFS,None,None,[],https://openalex.org/S7407052672,None,[],7407052672,None
1,DiSSCo,None,None,[],https://openalex.org/S7407053449,None,[],7407053449,None
2,Zenodo (CERN European Organization for Nuclear...,https://openalex.org/I67311998,European Organization for Nuclear Research,[https://openalex.org/I67311998],https://openalex.org/S4306400562,None,None,4306400562,https://www.wikidata.org/wiki/Q22661177
3,Figshare,https://openalex.org/I4210132348,Figshare (United Kingdom),[https://openalex.org/I4210132348],https://openalex.org/S4377196282,None,None,4377196282,None
4,PubMed,https://openalex.org/I1299303238,National Institutes of Health,[https://openalex.org/I1299303238],https://openalex.org/S4306525036,None,None,4306525036,https://www.wikidata.org/entity/Q180686


In [10]:
oa.head().isna().sum()

display_name                 0
host_organization            2
host_organization_name       2
host_organization_lineage    0
id_openalex                  0
id_issn_l                    5
id_issn                      3
id_mag                       0
id_wikidata                  3
dtype: int64

--> replace empty lists with None

In [11]:
oa.id_issn = oa.id_issn.apply(lambda x: None if (x is None or len(x) == 0) else x)
oa.host_organization_lineage = oa.host_organization_lineage.apply(lambda x: None if (x is None or len(x) == 0) else x)

In [12]:
oa.head()

,display_name,host_organization,host_organization_name,host_organization_lineage,id_openalex,id_issn_l,id_issn,id_mag,id_wikidata
0,NIFS,None,None,None,https://openalex.org/S7407052672,None,None,7407052672,None
1,DiSSCo,None,None,None,https://openalex.org/S7407053449,None,None,7407053449,None
2,Zenodo (CERN European Organization for Nuclear...,https://openalex.org/I67311998,European Organization for Nuclear Research,[https://openalex.org/I67311998],https://openalex.org/S4306400562,None,None,4306400562,https://www.wikidata.org/wiki/Q22661177
3,Figshare,https://openalex.org/I4210132348,Figshare (United Kingdom),[https://openalex.org/I4210132348],https://openalex.org/S4377196282,None,None,4377196282,None
4,PubMed,https://openalex.org/I1299303238,National Institutes of Health,[https://openalex.org/I1299303238],https://openalex.org/S4306525036,None,None,4306525036,https://www.wikidata.org/entity/Q180686


Is `id_mag` always same as `id_openalex`?

In [13]:
len("https://openalex.org/S")

22

In [14]:
(oa.id_openalex.str[22:] == oa.id_mag).all()

True

Yes, it both columns refer to the unique ID of the OpenAlex. However, in **Solr** `source_id`s do not always match with `id_mag`. 

### Identify ID columns for Mapping

Let's check some descriptive stats and identify columns that we can use to map with Scopus / Web of Science indeces.

In [15]:
oa.count()

display_name                 283280
host_organization             73287
host_organization_name        73510
host_organization_lineage     73743
id_openalex                  283281
id_issn_l                    166455
id_issn                      166455
id_mag                       283281
id_wikidata                   82149
dtype: int64

In [16]:
oa[["display_name", "id_mag", "id_issn_l", "host_organization"]].nunique()

display_name         272268
id_mag               283281
id_issn_l            166270
host_organization     12172
dtype: int64

Can't use `id_issn_l` as an id because it's not unique.. let's see whether the concatenated `id_issn` is unique.

In [17]:
oa.id_issn.str.join(",").sample(10)

80406                    None
45369               2687-4792
109321    1269-3286,1279-8509
137080    0853-5205,2964-5808
2429                     None
131513    2171-956X,2174-1131
172265              1309-8659
36836                    None
174948              2775-085X
115496                   None
Name: id_issn, dtype: object

In [18]:
oa["id_issn_concat"] = oa.id_issn.str.join(",")
oa.id_issn_concat.nunique() # compared to count 166455

166451

In [19]:
print(f"id_issn_l = count: {oa.id_issn_l.count()}, nunique: {oa.id_issn_l.nunique()}, num duplicates: {oa.id_issn_l.count() - oa.id_issn_l.nunique()}, nulls: {oa.id_issn_l.isna().sum()}")
print(f"id_issn   = count: {oa.id_issn.count()}, nunique: {oa.id_issn_concat.nunique()}, num duplicates:   {oa.id_issn.count() - oa.id_issn_concat.nunique()}, nulls: {oa.id_issn.isna().sum()}")

id_issn_l = count: 166455, nunique: 166270, num duplicates: 185, nulls: 116826
id_issn   = count: 166455, nunique: 166451, num duplicates:   4, nulls: 116826


In [20]:
# percentage missing data
oa.isna().sum() / oa.shape[0]

display_name                 0.000004
host_organization            0.741292
host_organization_name       0.740505
host_organization_lineage    0.739683
id_openalex                  0.000000
id_issn_l                    0.412403
id_issn                      0.412403
id_mag                       0.000000
id_wikidata                  0.710009
id_issn_concat               0.412403
dtype: float64

In [21]:
mask = oa.id_issn_concat.value_counts()[oa.id_issn_concat.value_counts() > 1]
oa.loc[(oa.id_issn_concat.isin(mask.index))].sort_values("id_issn_concat")

,display_name,host_organization,host_organization_name,host_organization_lineage,id_openalex,id_issn_l,id_issn,id_mag,id_wikidata,id_issn_concat
39428,Jahrbuch der Österreichischen Byzantinistik,None,None,None,https://openalex.org/S5406935911,0378-8660,"[0378-8660, 1810-536X]",5406935911,None,"0378-8660,1810-536X"
39685,Jahrbuch der Österreichischen Byzantinistik,None,None,[None],https://openalex.org/S4306514788,0378-8660,"[0378-8660, 1810-536X]",4306514788,http://www.wikidata.org/entity/Q1453509,"0378-8660,1810-536X"
214109,Revista Andina de Estudios Políticos,None,None,None,https://openalex.org/S5406983929,2221-4135,[2221-4135],5406983929,None,2221-4135
238697,Revista andina de estudios políticos.,None,Observatorio de la Región Andina,[None],https://openalex.org/S4404663026,2221-4135,[2221-4135],4404663026,None,2221-4135
40587,Akofena,None,None,None,https://openalex.org/S7407051855,2706-6312,"[2706-6312, 2708-0633]",7407051855,None,"2706-6312,2708-0633"
165231,Akofena revue scientifique des Sciences du Lan...,None,None,None,https://openalex.org/S7407056122,2706-6312,"[2706-6312, 2708-0633]",7407056122,None,"2706-6312,2708-0633"
223013,JSSAL,None,Chemnitz University of Technology,[None],https://openalex.org/S7407051706,2747-5646,[2747-5646],7407051706,None,2747-5646
255231,Journal of Research in Social Sciences and Lan...,None,None,None,https://openalex.org/S5407053510,2747-5646,[2747-5646],5407053510,None,2747-5646


Based on above, it looks like id_issn_l might be the latest ISSN. And the docs say it is "The ISSN-L (linking ISSN) for this source. This is the Canonical External ID for sources." 

So maybe we can use this column to map indeed. However first we need to understand why it has so many more duplicates than the list of all ISSNs used by the source.

In [22]:
mask = oa.id_issn_l.value_counts()[oa.id_issn_l.value_counts() > 1]
mask

id_issn_l
3108-3943    2
3090-5656    2
2714-2612    2
1846-1867    2
3110-0155    2
            ..
3117-3926    2
2523-9732    2
2833-1826    2
3079-9228    2
2981-8230    2
Name: count, Length: 185, dtype: int64

In [23]:
oa.loc[(oa.id_issn_l.isin(mask.index))].sort_values("id_issn_l").head(24)

,display_name,host_organization,host_organization_name,host_organization_lineage,id_openalex,id_issn_l,id_issn,id_mag,id_wikidata,id_issn_concat
232120,Listy filologické = Folia Philologica,None,None,None,https://openalex.org/S5407053357,0024-4457,[0024-4457],5407053357,None,0024-4457
232286,Listy filologické = Folia Philologica,None,None,None,https://openalex.org/S7407056394,0024-4457,"[0024-4457, 2570-9410]",7407056394,None,"0024-4457,2570-9410"
261698,Offa,None,Nomos Verlag,[None],https://openalex.org/S5406927939,0078-3714,[0078-3714],5406927939,None,0078-3714
189913,Offa,None,None,None,https://openalex.org/S5407054541,0078-3714,"[0078-3714, 2510-7933]",5407054541,None,"0078-3714,2510-7933"
162950,Šport revija za teoretična in praktična vpraša...,None,None,None,https://openalex.org/S7407058129,0353-7455,"[0353-7455, 3024-0603]",7407058129,None,"0353-7455,3024-0603"
255017,Šport revija za teoretična in praktična vpraša...,None,None,None,https://openalex.org/S5406934987,0353-7455,[0353-7455],5406934987,None,0353-7455
39428,Jahrbuch der Österreichischen Byzantinistik,None,None,None,https://openalex.org/S5406935911,0378-8660,"[0378-8660, 1810-536X]",5406935911,None,"0378-8660,1810-536X"
39685,Jahrbuch der Österreichischen Byzantinistik,None,None,[None],https://openalex.org/S4306514788,0378-8660,"[0378-8660, 1810-536X]",4306514788,http://www.wikidata.org/entity/Q1453509,"0378-8660,1810-536X"
116883,Prometheus Rivista di studi classici,None,None,None,https://openalex.org/S5407052210,0391-2698,[0391-2698],5407052210,None,0391-2698
116891,Prometheus Rivista di studi classici,None,None,None,https://openalex.org/S7407057212,0391-2698,"[0391-2698, 2281-1044]",7407057212,None,"0391-2698,2281-1044"


Looks like all duplications result from an update, and we can deduplicate by keeping the most information:
- if id_wikidata is not empty, choose the value from non-empty version
- if id_issn is longer, keep the longer version
- if host organisation name is not None, keep that one
- if host_organization_lineage is [None], this seems like a more recent way of updating data

From what I can think, the only reason why we would not want to keep the latest row would be needing the additional information to fall back to when we cannot match based on id_issn_l. Therefore before writing a special deduplication that always keeps the more information, rather than what OpenAlex believes to be currently correct - latest -, let's see how much we can match with just the ID.

### Deduplicate Open Alex Data

#### with last row

In [24]:
print(oa.shape)
oa = oa.groupby("id_issn_l").last().reset_index()
oa.shape

(283281, 10)


(166270, 10)

In [25]:
scie = pd.read_csv("../data/journal_indices/webofsci/Science Citation Index Expanded (SCIE).csv")
print(scie.shape)
scie.head()

(9430, 7)


,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories
0,2D MATERIALS,2053-1583,2053-1583,IOP PUBLISHING LTD,"No.2 The Distillery, Glassfields, Avon Street,...",English,Nanoscience & Nanotechnology | Materials Scien...
1,3 BIOTECH,2190-572X,2190-5738,SPRINGER HEIDELBERG,"TIERGARTENSTRASSE 17, HEIDELBERG, GERMANY, D-6...",English,Biotechnology & Applied Microbiology
2,3D PRINTING AND ADDITIVE MANUFACTURING,2329-7662,2329-7670,SAGE PUBLICATIONS INC,"2455 TELLER RD, THOUSAND OAKS, USA, CA, 91320",English,"Engineering, Manufacturing | Materials Science..."
3,4OR-A QUARTERLY JOURNAL OF OPERATIONS RESEARCH,1619-4500,1614-2411,SPRINGER HEIDELBERG,"TIERGARTENSTRASSE 17, HEIDELBERG, GERMANY, D-6...",English,Operations Research & Management Science
4,AAPG BULLETIN,0149-1423,1558-9153,AMER ASSOC PETROLEUM GEOLOGIST,"1444 S BOULDER AVE, PO BOX 979, TULSA, USA, OK...",English,"Geosciences, Multidisciplinary"


### Match coverage

In [26]:
scie2 = scie.merge(
    oa[["id_issn_l", "display_name", "host_organization", "host_organization_name", "id_mag"]], 
    how="left", left_on="ISSN", right_on="id_issn_l")
print(scie2.shape)
scie2.head()

(9430, 12)


,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories,id_issn_l,display_name,host_organization,host_organization_name,id_mag
0,2D MATERIALS,2053-1583,2053-1583,IOP PUBLISHING LTD,"No.2 The Distillery, Glassfields, Avon Street,...",English,Nanoscience & Nanotechnology | Materials Scien...,2053-1583,2D Materials,https://openalex.org/P4310320083,IOP Publishing,4210229423
1,3 BIOTECH,2190-572X,2190-5738,SPRINGER HEIDELBERG,"TIERGARTENSTRASSE 17, HEIDELBERG, GERMANY, D-6...",English,Biotechnology & Applied Microbiology,2190-572X,3 Biotech,https://openalex.org/P4310319900,Springer Science+Business Media,4210190239
2,3D PRINTING AND ADDITIVE MANUFACTURING,2329-7662,2329-7670,SAGE PUBLICATIONS INC,"2455 TELLER RD, THOUSAND OAKS, USA, CA, 91320",English,"Engineering, Manufacturing | Materials Science...",2329-7662,3D Printing and Additive Manufacturing,https://openalex.org/P4310320443,"Mary Ann Liebert, Inc.",4210174904
3,4OR-A QUARTERLY JOURNAL OF OPERATIONS RESEARCH,1619-4500,1614-2411,SPRINGER HEIDELBERG,"TIERGARTENSTRASSE 17, HEIDELBERG, GERMANY, D-6...",English,Operations Research & Management Science,NaN,NaN,NaN,NaN,NaN
4,AAPG BULLETIN,0149-1423,1558-9153,AMER ASSOC PETROLEUM GEOLOGIST,"1444 S BOULDER AVE, PO BOX 979, TULSA, USA, OK...",English,"Geosciences, Multidisciplinary",0149-1423,AAPG Bulletin,https://openalex.org/P4310316082,American Association of Petroleum Geologists,144020083


In [27]:
scie2.isna().sum() / scie2.shape[0]

Journal title                0.000000
ISSN                         0.060976
eISSN                        0.011877
Publisher name               0.000000
Publisher address            0.000000
Languages                    0.006045
Web of Science Categories    0.000212
id_issn_l                    0.108165
display_name                 0.108165
host_organization            0.151432
host_organization_name       0.151432
id_mag                       0.108165
dtype: float64

In [28]:
scie2[scie2.ISSN.isna()].sample(5)

,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories,id_issn_l,display_name,host_organization,host_organization_name,id_mag
8897,SYSTEMATIC ENTOMOLOGY,NaN,1365-3113,WILEY,"111 RIVER ST, HOBOKEN, USA, NJ, 07030-5774",English,Evolutionary Biology | Entomology,NaN,NaN,NaN,NaN,NaN
2714,EFSA JOURNAL,NaN,1831-4732,WILEY-V C H VERLAG GMBH,"POSTFACH 101161, WEINHEIM, GERMANY, 69451",English,Food Science & Technology,NaN,NaN,NaN,NaN,NaN
2465,DARU-JOURNAL OF PHARMACEUTICAL SCIENCES,NaN,2008-2231,SPRINGER INT PUBL AG,"GEWERBESTRASSE 11, CHAM, SWITZERLAND, CH-6330",English,Pharmacology & Pharmacy,NaN,NaN,NaN,NaN,NaN
2657,EARTHS FUTURE,NaN,2328-4277,AMER GEOPHYSICAL UNION,"2000 FLORIDA AVE NW, WASHINGTON, USA, DC, 20009",English,"Environmental Sciences | Geosciences, Multidis...",NaN,NaN,NaN,NaN,NaN
8059,PROGRESS OF THEORETICAL AND EXPERIMENTAL PHYSICS,NaN,2050-3911,OXFORD UNIV PRESS INC,"JOURNALS DEPT, 2001 EVANS RD, CARY, USA, NC, 2...",English,"Physics, Multidisciplinary | Physics, Particle...",NaN,NaN,NaN,NaN,NaN


In [29]:
scie3 = scie[scie.ISSN.isna()]
print(scie3.shape)
scie3 = scie3.merge(
    oa[["id_issn_l", "display_name", "host_organization", "host_organization_name", "id_mag"]], 
    how="left", left_on="eISSN", right_on="id_issn_l", validate="1:1")
print(scie3.shape)
scie3.isna().sum() / scie3.shape[0]

(575, 7)
(575, 12)


Journal title                0.000000
ISSN                         1.000000
eISSN                        0.000000
Publisher name               0.000000
Publisher address            0.000000
Languages                    0.046957
Web of Science Categories    0.003478
id_issn_l                    0.142609
display_name                 0.142609
host_organization            0.200000
host_organization_name       0.200000
id_mag                       0.142609
dtype: float64

In [30]:
scie3.sample(5)

,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories,id_issn_l,display_name,host_organization,host_organization_name,id_mag
441,NJAS-IMPACT IN AGRICULTURAL AND LIFE SCIENCES,NaN,2768-5241,TAYLOR & FRANCIS LTD,"2-4 PARK SQUARE, MILTON PARK, ABINGDON, Englan...",English,"Agriculture, Multidisciplinary",2768-5241,NJAS Impact in Agricultural and Life Sciences,https://openalex.org/P4310320547,Taylor & Francis,3207389948
411,MOLECULAR ECOLOGY,NaN,1365-294X,WILEY,"111 RIVER ST, HOBOKEN, USA, NJ, 07030-5774",English,Biochemistry & Molecular Biology | Ecology | E...,NaN,NaN,NaN,NaN,NaN
563,VETERINARY MEDICINE AND SCIENCE,NaN,2053-1095,WILEY,"111 RIVER ST, HOBOKEN, USA, NJ, 07030",English,Veterinary Sciences,2053-1095,Veterinary Medicine and Science,https://openalex.org/P4310320595,Wiley,4210212351
504,PSYCHIATRY INVESTIGATION,NaN,1976-3026,KOREAN NEUROPSYCHIATRIC ASSOC,"RN 522, G-FIVE CENTRAL PLAZA 1685-8 SEOCHO 4-D...",English,Psychiatry,NaN,NaN,NaN,NaN,NaN
231,FRONTIERS IN CELLULAR NEUROSCIENCE,NaN,1662-5102,FRONTIERS MEDIA SA,"AVENUE DU TRIBUNAL FEDERAL 34, LAUSANNE, SWITZ...",English,Neurosciences,1662-5102,Frontiers in Cellular Neuroscience,https://openalex.org/P4310320527,Frontiers Media,36653316


Names of Journals seem to match.. Let's fill the rest with string matching

In [31]:
scie4 = scie3[scie3.id_issn_l.isna()][['Journal title', 'ISSN', 'eISSN', 'Publisher name', 'Publisher address',
       'Languages', 'Web of Science Categories']]
scie4["journal_title"] = scie4["Journal title"].str.lower().str.strip()
scie4.head()

,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories,journal_title
5,ACTA BIOCHIMICA POLONICA,NaN,1734-154X,FRONTIERS MEDIA SA,"AVENUE DU TRIBUNAL FEDERAL 34, LAUSANNE, SWITZ...",English,Biochemistry & Molecular Biology,acta biochimica polonica
6,ACTA BOTANICA MEXICANA,NaN,2448-7589,INST ECOLOGIA AC,"CENTRO REGIONAL DEL BAJIO, APARTADO POSTAL 386...",Spanish,Plant Sciences,acta botanica mexicana
7,ACTA CRYSTALLOGRAPHICA SECTION B-STRUCTURAL SC...,NaN,2052-5206,INT UNION CRYSTALLOGRAPHY,"2 ABBEY SQ, CHESTER, ENGLAND, CH1 2HU",English,"Chemistry, Multidisciplinary | Crystallography",acta crystallographica section b-structural sc...
11,ADDICTION SCIENCE & CLINICAL PRACTICE,NaN,1940-0640,BMC,"CAMPUS, 4 CRINAN ST, LONDON, ENGLAND, N1 9XW",English,Substance Abuse,addiction science & clinical practice
20,AGRARFORSCHUNG SCHWEIZ,NaN,2813-317X,AGRARFORSCHUNG,"AGROSCOPE LIEBEFELD-POSIEUX, POSIEUX, SWITZERL...",Multi-Language,"Agriculture, Multidisciplinary",agrarforschung schweiz


In [32]:
oa["journal_title"] = oa["display_name"].str.lower().str.strip()
unique_title_mask = oa.journal_title.value_counts()[oa.journal_title.value_counts() == 1]
oa2 = oa.loc[(oa.journal_title.isin(unique_title_mask.index))]

print(scie4.shape)
scie4 = scie4.merge(
    oa2[["id_issn_l", "display_name", "host_organization", "host_organization_name", "id_mag", 
        "journal_title"]], 
    how="left", on="journal_title", validate="1:1")
print(scie4.shape)
scie4.isna().sum() / scie4.shape[0]

(82, 8)
(82, 13)


Journal title                0.000000
ISSN                         1.000000
eISSN                        0.000000
Publisher name               0.000000
Publisher address            0.000000
Languages                    0.060976
Web of Science Categories    0.012195
journal_title                0.000000
id_issn_l                    0.317073
display_name                 0.317073
host_organization            0.365854
host_organization_name       0.365854
id_mag                       0.317073
dtype: float64

In [33]:
oa.journal_title.value_counts().head()

journal_title
proceedings                            74
jurnal pengabdian kepada masyarakat    20
jurnal pengabdian masyarakat           15
j-institute                            14
jurnal kesehatan                       13
Name: count, dtype: int64

In [34]:
scie2.isna().sum() / scie2.shape[0]

Journal title                0.000000
ISSN                         0.060976
eISSN                        0.011877
Publisher name               0.000000
Publisher address            0.000000
Languages                    0.006045
Web of Science Categories    0.000212
id_issn_l                    0.108165
display_name                 0.108165
host_organization            0.151432
host_organization_name       0.151432
id_mag                       0.108165
dtype: float64

In [35]:
((scie2.ISSN.notna()) & (scie2.id_mag.isna())).sum()

445

In [36]:
scie2

,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories,id_issn_l,display_name,host_organization,host_organization_name,id_mag
0,2D MATERIALS,2053-1583,2053-1583,IOP PUBLISHING LTD,"No.2 The Distillery, Glassfields, Avon Street,...",English,Nanoscience & Nanotechnology | Materials Scien...,2053-1583,2D Materials,https://openalex.org/P4310320083,IOP Publishing,4210229423
1,3 BIOTECH,2190-572X,2190-5738,SPRINGER HEIDELBERG,"TIERGARTENSTRASSE 17, HEIDELBERG, GERMANY, D-6...",English,Biotechnology & Applied Microbiology,2190-572X,3 Biotech,https://openalex.org/P4310319900,Springer Science+Business Media,4210190239
2,3D PRINTING AND ADDITIVE MANUFACTURING,2329-7662,2329-7670,SAGE PUBLICATIONS INC,"2455 TELLER RD, THOUSAND OAKS, USA, CA, 91320",English,"Engineering, Manufacturing | Materials Science...",2329-7662,3D Printing and Additive Manufacturing,https://openalex.org/P4310320443,"Mary Ann Liebert, Inc.",4210174904
3,4OR-A QUARTERLY JOURNAL OF OPERATIONS RESEARCH,1619-4500,1614-2411,SPRINGER HEIDELBERG,"TIERGARTENSTRASSE 17, HEIDELBERG, GERMANY, D-6...",English,Operations Research & Management Science,NaN,NaN,NaN,NaN,NaN
4,AAPG BULLETIN,0149-1423,1558-9153,AMER ASSOC PETROLEUM GEOLOGIST,"1444 S BOULDER AVE, PO BOX 979, TULSA, USA, OK...",English,"Geosciences, Multidisciplinary",0149-1423,AAPG Bulletin,https://openalex.org/P4310316082,American Association of Petroleum Geologists,144020083
...,...,...,...,...,...,...,...,...,...,...,...,...
9425,ZOOSYSTEMA,1280-9551,1638-9387,"PUBLICATIONS SCIENTIFIQUES DU MUSEUM, PARIS","CP 39-57, RUE CUVIER, PARIS CEDEX 05, FRANCE, ...",English,Zoology,1280-9551,Zoosystema,https://openalex.org/P4310311095,Muséum national d'histoire naturelle,141665667
9426,ZOOSYSTEMATICS AND EVOLUTION,1860-0743,1860-0743,PENSOFT PUBLISHERS,"12 PROF GEORGI ZLATARSKI ST, SOFIA, BULGARIA, ...",Multi-Language,Zoology,NaN,NaN,NaN,NaN,NaN
9427,ZOOTAXA,1175-5326,1175-5334,MAGNOLIA PRESS,"250F Marua Road Mt Wellington, AUCKLAND, New Z...",English,Zoology,1175-5326,Zootaxa,https://openalex.org/P4310321855,Q15088586,171471881
9428,ZUCHTUNGSKUNDE,0044-5401,1867-4518,EUGEN ULMER GMBH CO,"POSTFACH 700561 WOLLGRASWEG 41, STUTTGART, GER...",German,"Agriculture, Dairy & Animal Science",0044-5401,Züchtungskunde,https://openalex.org/P4310319817,Eugen Ulmer Verlag,34199213


In [37]:
scie2[(scie2.ISSN.notna()) & (scie2.id_mag.isna())].count() / scie2.shape[0]

Journal title                0.047190
ISSN                         0.047190
eISSN                        0.046129
Publisher name               0.047190
Publisher address            0.047190
Languages                    0.046023
Web of Science Categories    0.047190
id_issn_l                    0.000000
display_name                 0.000000
host_organization            0.000000
host_organization_name       0.000000
id_mag                       0.000000
dtype: float64

with 2 falllbacks to eISSN and titles, we get only %30 of 82 rows unmatched --> 25 rows out of original 9430 - 445  --> 0.003 missing. Maybe that can be recovered with string beginning-ending matching..

In [38]:
scie4.sample(5)

,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories,journal_title,id_issn_l,display_name,host_organization,host_organization_name,id_mag
29,ETHICS AND SOCIETY,NaN,3091-3322,SPRINGER,"VAN GODEWIJCKSTRAAT 30, DORDRECHT, NETHERLANDS...",NaN,NaN,ethics and society,3091-3314,Ethics and Society,None,None,5407055850
2,ACTA CRYSTALLOGRAPHICA SECTION B-STRUCTURAL SC...,NaN,2052-5206,INT UNION CRYSTALLOGRAPHY,"2 ABBEY SQ, CHESTER, ENGLAND, CH1 2HU",English,"Chemistry, Multidisciplinary | Crystallography",acta crystallographica section b-structural sc...,NaN,NaN,NaN,NaN,NaN
43,INTERNATIONAL JOURNAL OF APPLIED CERAMIC TECHN...,NaN,1744-7402,WILEY,"111 RIVER ST, HOBOKEN, USA, NJ, 07030-5774",English,"Materials Science, Ceramics",international journal of applied ceramic techn...,1546-542X,International Journal of Applied Ceramic Techn...,https://openalex.org/P4310320595,Wiley,75568645
57,LANCET GASTROENTEROLOGY & HEPATOLOGY,NaN,2468-1253,ELSEVIER INC,"525 B STREET, STE 1900, SAN DIEGO, USA, CA, 92...",English,Gastroenterology & Hepatology,lancet gastroenterology & hepatology,NaN,NaN,NaN,NaN,NaN
60,MOLECULAR & CELLULAR PROTEOMICS,NaN,1535-9484,ELSEVIER,"RADARWEG 29, AMSTERDAM, Netherlands, 1043 NX",English,Biochemical Research Methods,molecular & cellular proteomics,1535-9476,Molecular & Cellular Proteomics,https://openalex.org/P4310320990,Elsevier BV,104550190


In [39]:
row = scie4[scie4.id_mag.isna()].sample()
title = row.journal_title.iloc[0]
print(row)
issn = row.ISSN.iloc[0]
if pd.isna(issn):
    issn = row.eISSN.iloc[0]
print(issn)
if pd.isna(issn):
    result = oa2[(oa2.journal_title.str.contains(title))]
else:
    result = oa2[(oa2.journal_title.str.contains(title)) | (oa2.id_issn_concat.str.contains(issn))]
result

                                    Journal title ISSN      eISSN  \
51  JOURNAL OF OTOLARYNGOLOGY-HEAD & NECK SURGERY  NaN  1916-0216   

           Publisher name                              Publisher address  \
51  SAGE PUBLICATIONS INC  2455 TELLER RD, THOUSAND OAKS, USA, CA, 91320   

   Languages Web of Science Categories  \
51   English       Otorhinolaryngology   

                                    journal_title id_issn_l display_name  \
51  journal of otolaryngology-head & neck surgery       NaN          NaN   

   host_organization host_organization_name id_mag  
51               NaN                    NaN    NaN  
1916-0216


,id_issn_l,display_name,host_organization,host_organization_name,host_organization_lineage,id_openalex,id_issn,id_mag,id_wikidata,id_issn_concat,journal_title
66512,1916-0208,Journal of Otolaryngology - Head and Neck Surgery,https://openalex.org/P4310320256,BioMed Central,"[https://openalex.org/P4310320256, https://ope...",https://openalex.org/S4210233012,"[1916-0208, 1916-0216]",4210233012,https://www.wikidata.org/entity/Q14633040,"1916-0208,1916-0216",journal of otolaryngology - head and neck surgery


In [40]:
title = "transfusion"
row = scie4[scie4.journal_title == title]
print(row)
issn = row.ISSN.iloc[0]
if pd.isna(issn):
    issn = row.eISSN.iloc[0]
print(issn)
if pd.isna(issn):
    result = oa2[(oa2.journal_title.str.contains(title))]
else:
    result = oa2[(oa2.journal_title.str.contains(title)) | (oa2.id_issn_concat.str.contains(issn))]
result

   Journal title ISSN      eISSN Publisher name  \
78   TRANSFUSION  NaN  1537-2995          WILEY   

                             Publisher address Languages  \
78  111 RIVER ST, HOBOKEN, USA, NJ, 07030-5774   English   

   Web of Science Categories journal_title id_issn_l display_name  \
78                Hematology   transfusion       NaN          NaN   

   host_organization host_organization_name id_mag  
78               NaN                    NaN    NaN  
1537-2995


,id_issn_l,display_name,host_organization,host_organization_name,host_organization_lineage,id_openalex,id_issn,id_mag,id_wikidata,id_issn_concat,journal_title
6578,0035-2977,Revue Française de Transfusion,https://openalex.org/P4310320990,Elsevier BV,[https://openalex.org/P4310320990],https://openalex.org/S4210180146,"[0035-2977, 2213-946X]",4210180146,None,"0035-2977,2213-946X",revue française de transfusion
9357,0074-8528,Proceedings of the International Society of Bl...,None,None,None,https://openalex.org/S4210225101,[0074-8528],4210225101,None,0074-8528,proceedings of the international society of bl...
14255,0171-7111,Haematology and blood transfusion,https://openalex.org/P4310319965,Springer Nature,[https://openalex.org/P4310319965],https://openalex.org/S2755861557,[0171-7111],2755861557,https://www.wikidata.org/entity/Q27719500,0171-7111,haematology and blood transfusion
18432,0278-6222,Plasma Therapy and Transfusion Technology,https://openalex.org/P4310320990,Elsevier BV,[https://openalex.org/P4310320990],https://openalex.org/S183557914,"[0278-6222, 1878-2019]",183557914,http://www.wikidata.org/entity/Q96328899,"0278-6222,1878-2019",plasma therapy and transfusion technology
18993,0301-3553,Ergebnisse der Bluttransfusionsforschung,None,None,None,https://openalex.org/S4210223732,[0301-3553],4210223732,http://www.wikidata.org/entity/Q93827960,0301-3553,ergebnisse der bluttransfusionsforschung
...,...,...,...,...,...,...,...,...,...,...,...
113314,2575-8934,Open Access Blood Research & Transfusion Journal,None,None,None,https://openalex.org/S4210221639,[2575-8934],4210221639,https://www.wikidata.org/entity/Q96721458,2575-8934,open access blood research & transfusion journal
149601,2978-8250,Transplant and Transfusion,None,None,None,https://openalex.org/S7407057746,[2978-8250],7407057746,None,2978-8250,transplant and transfusion
150012,2985-2390,Journal of Hematology and Transfusion Medicine,None,None,None,https://openalex.org/S4404677111,"[2985-2390, 2985-2404]",4404677111,None,"2985-2390,2985-2404",journal of hematology and transfusion medicine
161009,3070-1937,International Journal of Blood Transfusion,None,None,None,https://openalex.org/S5407054750,[3070-1937],5407054750,None,3070-1937,international journal of blood transfusion


!! with sting, we can accidentally match with something wrong.

#### Just with ID matching

So strategy should be
1) if ISSN, match with ISSN
2) if left has ISSN but right does not, match with other entry in ISSN list from right
3) left eISSn -> right issn_l
4) if left eISSN and right null -> check ISSN list from right

In [41]:
oa.shape

(166270, 11)

In [42]:
oa.head()

,id_issn_l,display_name,host_organization,host_organization_name,host_organization_lineage,id_openalex,id_issn,id_mag,id_wikidata,id_issn_concat,journal_title
0,,GEOTROPE,None,None,None,https://openalex.org/S7407051494,"[, 1817-5589, 2960-2084]",7407051494,None,",1817-5589,2960-2084",geotrope
1,0000-0205,Tropical Parasitology,https://openalex.org/P4310320448,Medknow,"[https://openalex.org/P4310320448, https://ope...",https://openalex.org/S4210234080,[0000-0205],4210234080,https://www.wikidata.org/entity/Q96738528,0000-0205,tropical parasitology
2,0000-1112,Journal of Musculoskeletal Diseases in Ayurveda,https://openalex.org/P4310320448,Medknow,"[https://openalex.org/P4310320448, https://ope...",https://openalex.org/S4387277910,[0000-1112],4387277910,None,0000-1112,journal of musculoskeletal diseases in ayurveda
3,0000-992X,Andrew Gilmartin Review,None,None,None,https://openalex.org/S5406922747,[0000-992X],5406922747,None,0000-992X,andrew gilmartin review
4,0001-0197,A A R N newsletter,None,None,None,https://openalex.org/S4306499931,[0001-0197],4306499931,http://www.wikidata.org/entity/Q27708670,0001-0197,a a r n newsletter


In [43]:
oa.id_issn.apply(len).value_counts()

id_issn
1    87643
2    77303
3     1207
4       96
5       17
7        2
6        2
Name: count, dtype: int64

In [44]:
oa.id_issn.head().apply(lambda x: [i for i in x if i]) #.apply(len).value_counts()

0    [1817-5589, 2960-2084]
1               [0000-0205]
2               [0000-1112]
3               [0000-992X]
4               [0001-0197]
Name: id_issn, dtype: object

In [45]:
oa.id_issn.apply(lambda x: [i for i in x if i]).apply(len).value_counts()

id_issn
1    87643
2    77304
3     1206
4       96
5       17
7        2
6        2
Name: count, dtype: int64

In [46]:
oa.sample(5)

,id_issn_l,display_name,host_organization,host_organization_name,host_organization_lineage,id_openalex,id_issn,id_mag,id_wikidata,id_issn_concat,journal_title
100994,2411-6734,ADBI policy brief,None,None,None,https://openalex.org/S4393918629,[2411-6734],4393918629,None,2411-6734,adbi policy brief
25973,0738-0151,Child and Adolescent Social Work Journal,https://openalex.org/P4310319900,Springer Science+Business Media,"[https://openalex.org/P4310319900, https://ope...",https://openalex.org/S4210168042,"[0738-0151, 1573-2797]",4210168042,None,"0738-0151,1573-2797",child and adolescent social work journal
131700,2721-3706,Jurnal Sinar Edukasi,None,None,None,https://openalex.org/S4387292681,"[2721-3706, 2721-6705]",4387292681,None,"2721-3706,2721-6705",jurnal sinar edukasi
83519,2198-350X,Mathematics for industry,https://openalex.org/P4310319965,Springer Nature,[https://openalex.org/P4310319965],https://openalex.org/S4210168673,"[2198-350X, 2198-3518]",4210168673,None,"2198-350X,2198-3518",mathematics for industry
119833,2622-6367,IQTISHADequity jurnal MANAJEMEN,None,None,None,https://openalex.org/S4210220197,[2622-6367],4210220197,None,2622-6367,iqtishadequity jurnal manajemen


Is `id_issn_l` alwyays the first in `id_issn` list? --> yes.

In [47]:
(oa.id_issn_l == oa.id_issn.apply(lambda x: x[0])).sum() / oa.shape[0]

1.0

In [48]:
scie["ISSN_all"] = scie[["ISSN", "eISSN"]].apply(lambda r: [v for v in r.tolist() if pd.notna(v)], axis=1)
scie["ISSN_filled"] = scie.ISSN
scie.loc[scie.ISSN_filled.isna(), "ISSN_filled"] = scie.loc[scie.ISSN_filled.isna(), "eISSN"]

scie.head()

,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories,ISSN_all,ISSN_filled
0,2D MATERIALS,2053-1583,2053-1583,IOP PUBLISHING LTD,"No.2 The Distillery, Glassfields, Avon Street,...",English,Nanoscience & Nanotechnology | Materials Scien...,"[2053-1583, 2053-1583]",2053-1583
1,3 BIOTECH,2190-572X,2190-5738,SPRINGER HEIDELBERG,"TIERGARTENSTRASSE 17, HEIDELBERG, GERMANY, D-6...",English,Biotechnology & Applied Microbiology,"[2190-572X, 2190-5738]",2190-572X
2,3D PRINTING AND ADDITIVE MANUFACTURING,2329-7662,2329-7670,SAGE PUBLICATIONS INC,"2455 TELLER RD, THOUSAND OAKS, USA, CA, 91320",English,"Engineering, Manufacturing | Materials Science...","[2329-7662, 2329-7670]",2329-7662
3,4OR-A QUARTERLY JOURNAL OF OPERATIONS RESEARCH,1619-4500,1614-2411,SPRINGER HEIDELBERG,"TIERGARTENSTRASSE 17, HEIDELBERG, GERMANY, D-6...",English,Operations Research & Management Science,"[1619-4500, 1614-2411]",1619-4500
4,AAPG BULLETIN,0149-1423,1558-9153,AMER ASSOC PETROLEUM GEOLOGIST,"1444 S BOULDER AVE, PO BOX 979, TULSA, USA, OK...",English,"Geosciences, Multidisciplinary","[0149-1423, 1558-9153]",0149-1423


In [58]:
def merge_by_bidirectional_preference(dfL, dfR,
                                      left_id_col='left_id', left_list_col='left_keys',
                                      right_id_col='right_id', right_list_col='right_keys',
                                      right_keep_cols=None,
                                      score_weights=(1.0, 1.0),
                                      ):
    """
    Merge dfL and dfR when both sides have ordered lists of alternative keys.
    Returns (result_df, stats) where result_df has one best match per left row (or NaNs if no match),
    and stats is a Series counting matches by chosen left_pref (l_pref) and right_pref (r_pref).

    Parameters:
    - dfL: left DataFrame. Must contain left_id_col and left_list_col (list-like per row).
    - dfR: right DataFrame. Must contain right_id_col and right_list_col (list-like per row).
    - left_id_col/right_id_col: identifier column names for left/right (must be present).
    - left_list_col/right_list_col: names of columns containing lists of alternatives.
    - right_keep_cols: list of columns from dfR to include in result (None => all except right_list_col).
    - score_weights: (wL, wR) weights for scoring: score = wL*l_pref + wR*r_pref.
    """
    tie_breaker=('score','l_pref','r_pref', right_id_col)

    # --- Prepare left exploded ---
    dfL_e = dfL.reset_index(drop=False).rename(columns={'index':'_orig_index'})  # preserve insertion order
    if left_id_col not in dfL_e:
        dfL_e[left_id_col] = dfL_e['_orig_index']
    dfL_e = dfL_e.explode(left_list_col).reset_index(drop=True)
    dfL_e['_lidx'] = dfL_e.groupby('_orig_index').cumcount()  # l_pref
    dfL_e = dfL_e.rename(columns={left_list_col: '_key_value'})
    dfL_e = dfL_e.rename(columns={'_lidx':'l_pref'})

    # --- Prepare right exploded ---
    dfR_e = dfR.reset_index(drop=False).rename(columns={'index':'_r_orig_index'})
    if right_id_col not in dfR_e:
        dfR_e[right_id_col] = dfR_e['_r_orig_index']
    dfR_e = dfR_e.explode(right_list_col).reset_index(drop=True)
    dfR_e['_ridx'] = dfR_e.groupby('_r_orig_index').cumcount()  # r_pref
    dfR_e = dfR_e.rename(columns={right_list_col: '_key_value', '_ridx':'r_pref'})

    # Which right columns to keep
    if right_keep_cols is None:
        right_keep_cols = [c for c in dfR_e.columns if c not in ('_key_value','r_pref','_r_orig_index')]
    right_keep_cols = list(dict.fromkeys(right_keep_cols))  # unique preserve order

    # --- Merge on key value ---
    merged = dfL_e.merge(dfR_e[[ '_key_value','r_pref','_r_orig_index'] + right_keep_cols ],
                         on='_key_value', how='left', suffixes=('_L','_R'))

    # --- Score & tie-break ---
    wL, wR = score_weights
    merged['l_pref'] = merged['l_pref'].astype(float)
    merged['r_pref'] = merged['r_pref'].astype(float)
    merged['score'] = wL * merged['l_pref'] + wR * merged['r_pref']

    # Sort for deterministic choice
    sort_cols = list(tie_breaker)
    # Ensure all tie_breaker cols exist in merged; if not, fallback to score,l_pref,r_pref
    for c in sort_cols:
        if c not in merged.columns:
            sort_cols = ['score','l_pref','r_pref']
            break
    merged_sorted = merged.sort_values(by=sort_cols)

    # Pick best per original left row (use _orig_index to map back)
    best = merged_sorted.drop_duplicates(subset=['_orig_index'], keep='first')

    # Build result: left original columns + chosen right columns + matching metadata
    left_cols = [c for c in dfL.columns if c not in (left_list_col,)]
    out_cols = left_cols + right_keep_cols + ['l_pref','r_pref','score','_key_value']
    result = best.reindex(columns=out_cols).reset_index(drop=True)

    # Stats: counts by l_pref and r_pref where match exists
    stats = {
        'total_left_rows': dfL.shape[0],
        'matched_rows': int(result[right_id_col].notna().sum()),
        'unmatched_rows': int(result[right_id_col].isna().sum())
    }
    # distribution by l_pref and r_pref
    stats['matched_by_l_pref'] = result.loc[result[right_id_col].notna(), 'l_pref'].value_counts().sort_index()
    stats['matched_by_r_pref'] = result.loc[result[right_id_col].notna(), 'r_pref'].value_counts().sort_index()

    return result, stats

In [59]:
print("Left shape:  ", scie.shape)
print("Right shape: ", oa.shape)
m, s = merge_by_bidirectional_preference(scie, oa, left_id_col='ISSN_filled', left_list_col='ISSN_all',
                                      right_id_col='id_issn_l', right_list_col='id_issn')
print("Merged shape:", m.shape)
m.isna().sum() / m.shape[0]

Left shape:   (9430, 9)
Right shape:  (166270, 11)
Merged shape: (9430, 22)


Journal title                0.000000
ISSN                         0.060976
eISSN                        0.011877
Publisher name               0.000000
Publisher address            0.000000
Languages                    0.006045
Web of Science Categories    0.000212
ISSN_filled                  0.000000
id_issn_l                    0.002651
display_name                 0.002651
host_organization            0.054295
host_organization_name       0.054295
host_organization_lineage    0.054189
id_openalex                  0.002651
id_mag                       0.002651
id_wikidata                  0.195758
id_issn_concat               0.002651
journal_title                0.002651
l_pref                       0.000000
r_pref                       0.002651
score                        0.002651
_key_value                   0.000000
dtype: float64

In [60]:
s

{'total_left_rows': 9430,
 'matched_rows': 9405,
 'unmatched_rows': 25,
 'matched_by_l_pref': l_pref
 0.0    9371
 1.0      34
 Name: count, dtype: int64,
 'matched_by_r_pref': r_pref
 0.0    8935
 1.0     447
 2.0      23
 Name: count, dtype: int64}

In [61]:
scie.merge(oa, how='left', left_on='ISSN_filled', right_on='id_issn_l').isna().sum()

Journal title                   0
ISSN                          575
eISSN                         112
Publisher name                  0
Publisher address               0
Languages                      57
Web of Science Categories       2
ISSN_all                        0
ISSN_filled                     0
id_issn_l                     527
display_name                  527
host_organization             968
host_organization_name        968
host_organization_lineage     967
id_openalex                   527
id_issn                       527
id_mag                        527
id_wikidata                  2212
id_issn_concat                527
journal_title                 527
dtype: int64

In [62]:
m.isna().sum()

Journal title                   0
ISSN                          575
eISSN                         112
Publisher name                  0
Publisher address               0
Languages                      57
Web of Science Categories       2
ISSN_filled                     0
id_issn_l                      25
display_name                   25
host_organization             512
host_organization_name        512
host_organization_lineage     511
id_openalex                    25
id_mag                         25
id_wikidata                  1846
id_issn_concat                 25
journal_title                  25
l_pref                          0
r_pref                         25
score                          25
_key_value                      0
dtype: int64

In [63]:
m.head()

,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories,ISSN_filled,id_issn_l,display_name,host_organization,host_organization_name,host_organization_lineage,id_openalex,id_mag,id_wikidata,id_issn_concat,journal_title,l_pref,r_pref,score,_key_value
0,COMMUNICATIONS OF THE ACM,0001-0782,1557-7317,ASSOC COMPUTING MACHINERY,"1601 Broadway, 10th Floor, NEW YORK, Usa, NY, ...",English,"Computer Science, Hardware & Architecture | Co...",0001-0782,0001-0782,Communications of the ACM,https://openalex.org/P4310319798,Association for Computing Machinery,[https://openalex.org/P4310319798],https://openalex.org/S103482838,103482838,https://www.wikidata.org/entity/Q1120519,"0001-0782,1557-7317",communications of the acm,0.0,0.0,0.0,0001-0782
1,AIAA JOURNAL,0001-1452,1533-385X,AMER INST AERONAUTICS ASTRONAUTICS,"1801 ALEXANDER BELL DRIVE, STE 500, RESTON, US...",English,"Engineering, Aerospace",0001-1452,0001-1452,AIAA Journal,https://openalex.org/P4310315709,American Institute of Aeronautics and Astronau...,[https://openalex.org/P4310315709],https://openalex.org/S53336109,53336109,https://www.wikidata.org/entity/Q4651795,"0001-1452,1081-0102,1533-385X",aiaa journal,0.0,0.0,0.0,0001-1452
2,AICHE JOURNAL,0001-1541,1547-5905,WILEY,"111 RIVER ST, HOBOKEN, USA, NJ, 07030-5774",English,"Engineering, Chemical",0001-1541,0001-1541,AIChE Journal,https://openalex.org/P4310320595,Wiley,[https://openalex.org/P4310320595],https://openalex.org/S147177414,147177414,https://www.wikidata.org/entity/Q15753322,"0001-1541,1547-5905",aiche journal,0.0,0.0,0.0,0001-1541
3,AORN JOURNAL,0001-2092,1878-0369,WILEY,"111 RIVER ST, HOBOKEN, USA, NJ, 07030-5774",English,Nursing,0001-2092,0001-2092,AORN Journal,https://openalex.org/P4310320595,Wiley,[https://openalex.org/P4310320595],https://openalex.org/S5303612,5303612,https://www.wikidata.org/entity/Q4653156,"0001-2092,1878-0369",aorn journal,0.0,0.0,0.0,0001-2092
4,ASHRAE JOURNAL,0001-2491,1943-6637,AMER SOC HEATING REFRIGERATING AIR-CONDITIONIN...,"1791 TULLIE CIRCLE NE, ATLANTA, USA, GA, 30329",English,Thermodynamics | Construction & Building Techn...,0001-2491,0001-2491,ASHRAE journal,https://openalex.org/P4310320990,Elsevier BV,[https://openalex.org/P4310320990],https://openalex.org/S1170834,1170834,https://www.wikidata.org/entity/Q15751177,0001-2491,ashrae journal,0.0,0.0,0.0,0001-2491


In [64]:
# sanity check
m[m.ISSN !=  m.id_issn_l][["ISSN", "eISSN", "_key_value", "id_issn_l", "id_issn_concat", "Journal title", "display_name" ]].dropna()

,ISSN,eISSN,_key_value,id_issn_l,id_issn_concat,Journal title,display_name
8904,0236-5294,1588-2632,0236-5294,0001-5954,"0001-5954,0236-5294,1588-2632",ACTA MATHEMATICA HUNGARICA,Acta Mathematica Academiae Scientiarum Hungaricae
8906,1612-9199,1612-9199,1612-9199,0003-9098,"0003-9098,1612-9199",EUROPEAN POULTRY SCIENCE,Archiv für Geflügelkunde
8907,1464-7931,1469-185X,1464-7931,0006-3231,"0006-3231,1464-7931,1469-185X",BIOLOGICAL REVIEWS,Biological reviews/Biological reviews of the C...
8908,1341-1357,1881-7122,1341-1357,0007-5124,"0007-5124,1341-1357,1881-7122,1884-4162,1884-4170",EXPERIMENTAL ANIMALS,EXPERIMENTAL ANIMALS
8909,2245-7070,2245-7070,2245-7070,0011-6297,"0011-6297,2245-7070",BULLETIN OF THE GEOLOGICAL SOCIETY OF DENMARK,Bulletin of the Geological Society of Denmark
...,...,...,...,...,...,...,...
9400,1735-0328,1726-6890,1735-0328,1726-6882,"1726-6882,1726-6890,1735-0328",IRANIAN JOURNAL OF PHARMACEUTICAL RESEARCH,Iranian journal of pharmaceutical research
9401,1749-4877,1749-4869,1749-4877,1749-4868,"1749-4868,1749-4869,1749-4877",INTEGRATIVE ZOOLOGY,Integrative Zoology
9402,2468-0133,2468-0133,2468-0133,2096-0522,"2096-0522,2096-675X,2468-0133",JOURNAL OF OCEAN ENGINEERING AND SCIENCE,Journal of Ocean Engineering and Science
9403,2225-6253,2411-9717,2411-9717,0038-223X,"0038-223X,2411-9717",JOURNAL OF THE SOUTHERN AFRICAN INSTITUTE OF M...,Journal of the Southern African Institute of M...


In [65]:
m[m.ISSN !=  m.id_issn_l][["ISSN", "eISSN", "_key_value", "id_issn_l", "id_issn_concat", "Journal title", "display_name" ]].dropna().sample(25)

,ISSN,eISSN,_key_value,id_issn_l,id_issn_concat,Journal title,display_name
9142,1742-7932,1742-7924,1742-7932,1742-7924,"1742-7924,1742-7932",JAPAN JOURNAL OF NURSING SCIENCE,Japan Journal of Nursing Science
9336,2724-5985,2724-5365,2724-5985,2724-5365,"2724-5365,2724-5985",MINERVA GASTROENTEROLOGY,Minerva Gastroenterology
9236,2058-5225,2058-1742,2058-5225,2058-1742,"2058-1742,2058-5225",EUROPEAN HEART JOURNAL-QUALITY OF CARE AND CLI...,European Heart Journal - Quality of Care and C...
9143,1743-9191,1743-9159,1743-9191,1743-9159,"1743-9159,1743-9191",INTERNATIONAL JOURNAL OF SURGERY,International Journal of Surgery
9352,1002-185X,1875-5372,1875-5372,1875-5372,1875-5372,RARE METAL MATERIALS AND ENGINEERING,Rare Metal Materials and Engineering
8961,0884-2914,2044-5326,0884-2914,0884-1616,"0884-1616,0884-2914,1091-8876,1092-8928,2044-5326",JOURNAL OF MATERIALS RESEARCH,Journal of materials research/Pratt's guide to...
9108,1598-8198,1598-818X,1598-8198,1598-818X,"1598-818X,1598-8198",COMPUTERS AND CONCRETE,"Computers and Concrete, an International Journal"
9190,1931-5244,1878-1810,1931-5244,1878-1810,"1878-1810,1931-5244",TRANSLATIONAL RESEARCH,Translational research
9266,2193-567X,2191-4281,2193-567X,2191-4281,"2191-4281,2193-567X",ARABIAN JOURNAL FOR SCIENCE AND ENGINEERING,Arabian Journal for Science and Engineering
8945,1600-0889,1600-0889,1600-0889,0280-6509,"0280-6509,1600-0889",TELLUS SERIES B-CHEMICAL AND PHYSICAL METEOROLOGY,Tellus B


ok, matching from older ISSNs in OpenAlex or switching to eISSN when ISSN is not found in web of science index seems to match reasonably based on the titles of the journals. We have only 0.002651 missing (25 rows). Let's check what's still missing for the last time.

In [66]:
m[m.id_issn_l.isna()]

,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories,ISSN_filled,id_issn_l,display_name,host_organization,host_organization_name,host_organization_lineage,id_openalex,id_mag,id_wikidata,id_issn_concat,journal_title,l_pref,r_pref,score,_key_value
9405,ACTA PHYTOTAXONOMICA ET GEOBOTANICA,1346-7565,2189-7042,"JAPANESE SOC PLANT SYSTEMATICS, UNIV TOKYO, GR...","BOTANICAL GARDENS, 3-7-1 HAKUSAN, TOKYO, JAPAN...",English,Plant Sciences,1346-7565,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,1346-7565
9406,AGRARFORSCHUNG SCHWEIZ,NaN,2813-317X,AGRARFORSCHUNG,"AGROSCOPE LIEBEFELD-POSIEUX, POSIEUX, SWITZERL...",Multi-Language,"Agriculture, Multidisciplinary",2813-317X,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,2813-317X
9407,ANNALI DI BOTANICA,0365-0812,2239-3129,UNIV STUDI ROMA LA SAPIENZA,"PIAZZALE ALDO MORO, 5, ROME, ITALY, I-00185",Multi-Language,Plant Sciences | Environmental Sciences,0365-0812,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,0365-0812
9408,ANNALS ACADEMY OF MEDICINE SINGAPORE,NaN,2972-4066,ACAD MEDICINE SINGAPORE,"81 Kim Keat Road, #12-00 NKF Centre, Singapore...",English,"Medicine, General & Internal",2972-4066,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,2972-4066
9409,ASIAN MYRMECOLOGY,1985-1944,NaN,UNIV MALAYSIA SABAH,"JALAN UMS, 88400, KOTA KINABALU, MALAYSIA, SAB...",English,Entomology,1985-1944,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,1985-1944
9410,ATW-INTERNATIONAL JOURNAL FOR NUCLEAR POWER,1431-5254,NaN,INFORUM VERLAGS-VERWALTUNGSGESELLSCHAFT MBH,"ATW-ADVERTISING-SUBSCRIPTIONS, ROBERT-KOCH-PLA...",Multi-Language,Nuclear Science & Technology,1431-5254,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,1431-5254
9411,AUSTRIAN JOURNAL OF FOREST SCIENCE,0379-5292,0375-524X,OSTERREICHISCHER AGRARVERLAG,"BANKGASSE 13, VIENNA 1, AUSTRIA, 1014",German,Forestry,0379-5292,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,0379-5292
9412,BERICHTE UBER LANDWIRTSCHAFT,2196-5099,NaN,BUNDESMINISTERIUM ERNAHRUNG LANDWIRTSCHAFT,"DEICHMANNS AUE 29, BPMM, GERMANY, 53179",Multi-Language,"Agriculture, Multidisciplinary",2196-5099,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,2196-5099
9413,BIOGEOCHEMICAL ADVANCES,NaN,3091-4493,SPRINGERNATURE,"CAMPUS, 4 CRINAN ST, LONDON, ENGLAND, N1 9XW",NaN,Geochemistry & Geophysics,3091-4493,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,3091-4493
9414,BULLETIN OF GEOPHYSICS AND OCEANOGRAPHY,2785-339X,2785-2970,IST NAZIONALE DI OCEANOGRAFIA E DI GEOFISICA,"BORGO GROTTA GIGANTE, 42-C, SGONICO, ITALY, TR...",English,Geochemistry & Geophysics,2785-339X,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,2785-339X


Let's see if we can get rid of anything with 100% title matching in the remaining missing 25 rows

In [68]:
m2 = m.loc[((m.id_issn_l.isna()) & (m._key_value.notna())), ['Journal title', 'ISSN', 'eISSN', 'Publisher name', 'Publisher address',
       'Languages', 'Web of Science Categories', 'ISSN_filled',]]
m2["journal_title"] = m2["Journal title"].str.lower().str.strip()
m2.merge(oa2, on="journal_title", how='left')

,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories,ISSN_filled,journal_title,id_issn_l,display_name,host_organization,host_organization_name,host_organization_lineage,id_openalex,id_issn,id_mag,id_wikidata,id_issn_concat
0,ACTA PHYTOTAXONOMICA ET GEOBOTANICA,1346-7565,2189-7042,"JAPANESE SOC PLANT SYSTEMATICS, UNIV TOKYO, GR...","BOTANICAL GARDENS, 3-7-1 HAKUSAN, TOKYO, JAPAN...",English,Plant Sciences,1346-7565,acta phytotaxonomica et geobotanica,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,AGRARFORSCHUNG SCHWEIZ,NaN,2813-317X,AGRARFORSCHUNG,"AGROSCOPE LIEBEFELD-POSIEUX, POSIEUX, SWITZERL...",Multi-Language,"Agriculture, Multidisciplinary",2813-317X,agrarforschung schweiz,1663-7852,Agrarforschung Schweiz,None,None,None,https://openalex.org/S2490815337,"[1663-7852, 1663-7909]",2490815337,http://www.wikidata.org/entity/Q96325599,"1663-7852,1663-7909"
2,ANNALI DI BOTANICA,0365-0812,2239-3129,UNIV STUDI ROMA LA SAPIENZA,"PIAZZALE ALDO MORO, 5, ROME, ITALY, I-00185",Multi-Language,Plant Sciences | Environmental Sciences,0365-0812,annali di botanica,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,ANNALS ACADEMY OF MEDICINE SINGAPORE,NaN,2972-4066,ACAD MEDICINE SINGAPORE,"81 Kim Keat Road, #12-00 NKF Centre, Singapore...",English,"Medicine, General & Internal",2972-4066,annals academy of medicine singapore,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,ASIAN MYRMECOLOGY,1985-1944,NaN,UNIV MALAYSIA SABAH,"JALAN UMS, 88400, KOTA KINABALU, MALAYSIA, SAB...",English,Entomology,1985-1944,asian myrmecology,2462-2362,Asian Myrmecology,https://openalex.org/P4310322526,Universiti Malaysia Sabah,[https://openalex.org/P4310322526],https://openalex.org/S2482358896,[2462-2362],2482358896,None,2462-2362
5,ATW-INTERNATIONAL JOURNAL FOR NUCLEAR POWER,1431-5254,NaN,INFORUM VERLAGS-VERWALTUNGSGESELLSCHAFT MBH,"ATW-ADVERTISING-SUBSCRIPTIONS, ROBERT-KOCH-PLA...",Multi-Language,Nuclear Science & Technology,1431-5254,atw-international journal for nuclear power,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
6,AUSTRIAN JOURNAL OF FOREST SCIENCE,0379-5292,0375-524X,OSTERREICHISCHER AGRARVERLAG,"BANKGASSE 13, VIENNA 1, AUSTRIA, 1014",German,Forestry,0379-5292,austrian journal of forest science,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7,BERICHTE UBER LANDWIRTSCHAFT,2196-5099,NaN,BUNDESMINISTERIUM ERNAHRUNG LANDWIRTSCHAFT,"DEICHMANNS AUE 29, BPMM, GERMANY, 53179",Multi-Language,"Agriculture, Multidisciplinary",2196-5099,berichte uber landwirtschaft,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
8,BIOGEOCHEMICAL ADVANCES,NaN,3091-4493,SPRINGERNATURE,"CAMPUS, 4 CRINAN ST, LONDON, ENGLAND, N1 9XW",NaN,Geochemistry & Geophysics,3091-4493,biogeochemical advances,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
9,BULLETIN OF GEOPHYSICS AND OCEANOGRAPHY,2785-339X,2785-2970,IST NAZIONALE DI OCEANOGRAFIA E DI GEOFISICA,"BORGO GROTTA GIGANTE, 42-C, SGONICO, ITALY, TR...",English,Geochemistry & Geophysics,2785-339X,bulletin of geophysics and oceanography,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [69]:
oa.loc[oa.id_mag=="2764689191", "host_organization_name"].iloc[0]

'Section Applied Botany of the German Botanical Society, German Society for Quality Research on Plant Foods (DGQ)'

we managed to match 2/19 with exact title match and one of them is wrong. so Unless we want to do it by hand, I wil leave these missing for now.

In [ ]:
merged_sorted = pd.read_csv("merged.csv")
#debug
d = merged_sorted[merged_sorted._key_value == "1928-9022"] # creating NaN output
d2 = merged_sorted[merged_sorted._key_value == "2196-5099"] # creating normal output
    
# Pick best per original left row (use _orig_index to map back)
best = merged_sorted.drop_duplicates(subset=['_orig_index'], keep='first')

: 

: 

In [ ]:
d

,Unnamed: 0,_orig_index,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories,_key_value,ISSN_filled,l_pref,r_pref,_r_orig_index,id_issn_l,display_name,host_organization,host_organization_name,host_organization_lineage,id_openalex,id_mag,id_wikidata,id_issn_concat,journal_title,score
18706,3237,1618,CANADIAN JOURNAL OF VETERINARY RESEARCH-REVUE ...,NaN,1928-9022,CANADIAN VET MED ASSOC,"339 BOOTH ST, OTTAWA, CANADA, ONTARIO, K1R 7K1",Multi-Language,Veterinary Sciences,1928-9022,1928-9022,1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


: 

: 

In [ ]:
d2

,Unnamed: 0,_orig_index,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories,_key_value,ISSN_filled,l_pref,r_pref,_r_orig_index,id_issn_l,display_name,host_organization,host_organization_name,host_organization_lineage,id_openalex,id_mag,id_wikidata,id_issn_concat,journal_title,score
18106,2266,1133,BERICHTE UBER LANDWIRTSCHAFT,2196-5099,NaN,BUNDESMINISTERIUM ERNAHRUNG LANDWIRTSCHAFT,"DEICHMANNS AUE 29, BPMM, GERMANY, 53179",Multi-Language,"Agriculture, Multidisciplinary",2196-5099,2196-5099,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


: 

: 

In [ ]:
merged_sorted[merged_sorted._orig_index==1618]

,Unnamed: 0,_orig_index,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories,_key_value,ISSN_filled,l_pref,r_pref,_r_orig_index,id_issn_l,display_name,host_organization,host_organization_name,host_organization_lineage,id_openalex,id_mag,id_wikidata,id_issn_concat,journal_title,score
18169,3236,1618,CANADIAN JOURNAL OF VETERINARY RESEARCH-REVUE ...,NaN,1928-9022,CANADIAN VET MED ASSOC,"339 BOOTH ST, OTTAWA, CANADA, ONTARIO, K1R 7K1",Multi-Language,Veterinary Sciences,NaN,1928-9022,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
18706,3237,1618,CANADIAN JOURNAL OF VETERINARY RESEARCH-REVUE ...,NaN,1928-9022,CANADIAN VET MED ASSOC,"339 BOOTH ST, OTTAWA, CANADA, ONTARIO, K1R 7K1",Multi-Language,Veterinary Sciences,1928-9022,1928-9022,1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


: 

: 

In [ ]:
scie[scie.ISSN_filled=="1928-9022"]

,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories,ISSN_all,ISSN_filled
1618,CANADIAN JOURNAL OF VETERINARY RESEARCH-REVUE ...,NaN,1928-9022,CANADIAN VET MED ASSOC,"339 BOOTH ST, OTTAWA, CANADA, ONTARIO, K1R 7K1",Multi-Language,Veterinary Sciences,"[nan, 1928-9022]",1928-9022


: 

: 

In [ ]:
pd.isna(scie[scie.ISSN_filled=="1928-9022"].ISSN_all.iloc[0][0])

True

: 

: 

In [ ]:
scie[scie.ISSN_filled=="1928-9022"] 2196-5099

,Journal title,ISSN,eISSN,Publisher name,Publisher address,Languages,Web of Science Categories,ISSN_all,ISSN_filled
1133,False,False,True,False,False,False,False,False,False


: 

: 